compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6038,6 +6040,25 @@ public void visit(BLangCheckedExpr checkedExpr, AnalyzerData data) {
         visitCheckAndCheckPanicExpr(checkedExpr, data);
     }
 
+    @Override
+    public void visit(BLangCheckedOnFailExpr checkedOnFailExpr, AnalyzerData data) {
+        BCheckedOnFailSymbol checkOnFailSymbol = new BCheckedOnFailSymbol(SymTag.CHECKED_ON_FAIL,
+                Flags.asMask(new HashSet<>(Lists.of())),
+                new Name(String.format("$checked_on_fail_symbol_%d$", data.commonAnalyzerData.letCount++)),
+                data.env.enclPkg.symbol.pkgID, checkedOnFailExpr.getBType(), data.env.scope.owner,
+                checkedOnFailExpr.pos);
+        data.env = SymbolEnv.createExprEnv(checkedOnFailExpr, data.env, checkOnFailSymbol);
+        setErrorVariableTypeInCheckedOnFailExpr(checkedOnFailExpr.simpleVariable, symTable.errorType, data);
+        checkExpr(checkedOnFailExpr.checkedExpr, data);
+        checkExpr(checkedOnFailExpr.expr, symTable.errorType, data);
```

Comment:
Now that the error constructor is not enforced by the parser, the different expressions give these two diagnostics.  
```bal
map<int> m = {a:1};

public function main() returns error? {
    int _ = check foo() on fail e => {a:1}; // Gives "a type compatible with mapping constructor expressions not found in type 'error'"
    int _ = check foo() on fail e => m; // Gives incompatible types: expected 'error', found 'map<int>'
}
```

Reply:
```
map<int> m = {a:1};

public function main() {
    error x = {a: 1}; // a type compatible with mapping constructor expressions not found in type 'error'
    error y = m; // Gives incompatible types: expected 'error', found 'map<int>'
}
```
This seems an existing issue. Not occurred by this change.